source/_posts/2019-10-07-upgrading-psr-interfaces.md

```diff
@@ -0,0 +1,67 @@
+
+PHP doesn't stand still, of course, and has added a host of **new features** in the last several years that could impact specifications. Most notable is the introduction of scalar types and return types in PHP 7.0, and of the `object`, `iterable`, and `void` types since then. (As of this writing it seems likely that we'll get union types in PHP 8, which is squeeeee!)
+
+It would be lovely to be able to include those new types in PHP-FIG interfaces, as that wouldn't change the semantics of the specs at all (they generally are already very specific about what parameter and return types they use) but would make writing a conformant implementation easier and make using a PSR-implementing library easier for everyone.
+
+The challenge, of course, is that **technically adding types to an interface is a backward-compatibility (BC) break**. That means doing so would immediately require implementers to make a BC break of their own to implement a new typed version of a PSR. While they certainly can, it also then means that **an end user then cannot use one library that has adopted the new typed version of a spec** and another that uses the old untyped version, because they're incompatible. And since they would share the same class name, a project could not have both installed at the same time. It's an all-or-nothing change, and those tend to be very not-fun. Which means they just wouldn't get used.
+
+### What won't work
+
+Ignoring the internal PHP-FIG process questions for now (as those are largely irrelevant to this discussion), there have been a few ways to go about upgrading specs that have largely been rejected over the years.
+
+1. **Just do it**: The sledgehammer approach, this would mean **ignoring the backward compatibility issues** and just releasing a 2.x tag of the interface specs with type hints and a higher PHP minimum version and calling it a day. While certainly the simplest for PHP-FIG, it's not the most viable for the community at large for the reasons described above.
+2. **Use an alternate namespace**: Another proposal has been to **version the namespace** for upgraded PSRs. That is, the Logger specification (PSR-3) currently uses the `\Psr\Log` namespace. So a fully type-enabled new version would be something like `\Psr\Log\V2`. That has the advantage that it would allow both the old and new version to be installed at the same time. However, it has the disadvantage that a library could not easily support both old and new at the same time, at least not without some tricky bridge-interface inheritance dance. It's doubtful that this would make the upgrade process any easier.
```

Comment:
```suggestion
2. **Use an alternate namespace**: Another proposal has been to **version the namespace** for upgraded PSRs. For example, the Logger specification (PSR-3) currently uses the `\Psr\Log` namespace. So a fully type-enabled new version would be something like `\Psr\Log\V2`. That has the advantage that it would allow both the old and new version to be installed at the same time. However, it has the disadvantage that a library could not easily support both old and new at the same time, at least not without some tricky bridge-interface inheritance dance. It's doubtful that this would make the upgrade process any easier.
```